src/luaotfload-harf-plug.lua

```diff
@@ -0,0 +1,1000 @@
+            for n, id, subtype in traverse(pre) do
+              precodes[#precodes + 1] = getfont(n) == fontid and getchar(n) or 0xFFFC
+            end
+            for n, id, subtype in traverse(post) do
+              postcodes[#postcodes + 1] = getfont(n) == fontid and getchar(n) or 0xFFFC
+            end
+            table.move(subcodes, subindex + 1, #subcodes, #postcodes + 1, postcodes)
+            do local newpre = copynodelist(startnode, disc)
+               setnext(tail(newpre), pre)
+               pre = newpre end
+            if post then
+              setnext(lastpost, copynodelist(getnext(disc), stopnode))
+            else
+              post = copynodelist(getnext(disc), stopnode)
+            end
+            if startnode ~= disc then
+              local predisc = getprev(disc)
+              setnext(predisc, rep)
+              setprev(rep, predisc)
+              if startnode == head then
+                head = disc
+              else
+                local before = getprev(startnode)
+                setnext(before, disc)
+                setprev(disc, before)
+              end
+              setprev(startnode, nil)
+              rep = startnode
+              lastrep = lastrep or predisc
+            end
+            if getnext(disc) ~= stopnode then
+              setnext(getprev(stopnode), nil)
+              setprev(stopnode, disc)
+              setprev(getnext(disc), lastrep)
+              setnext(lastrep, getnext(disc))
+              rep = rep or getnext(disc)
+              setnext(disc, stopnode)
+            end
+            glyph.replace = makesub(run, repcodes, rep)
+            glyph.pre = makesub(run, precodes, pre)
+            glyph.post = makesub(run, postcodes, post)
+            i = startglyph
+            node = disc
+          end
+        end
+      end
+      node = getnext(node)
+    end
+    codes.offset = run.len - len + (codes.offset or 0)
+    return head, glyphs
+  end
+
+  return head, {}
+end
+
+local function color_to_rgba(color)
+  local r = color.red   / 255
+  local g = color.green / 255
+  local b = color.blue  / 255
+  local a = color.alpha / 255
+  if a ~= 1 then
+    -- XXX: alpha
+    return format('%s %s %s rg', r, g, b)
+  else
+    return format('%s %s %s rg', r, g, b)
+  end
+end
+
+-- Cache of color glyph PNG data for bookkeeping, only because I couldn’t
+-- figure how to make the engine load the image from the binary data directly.
+local pngcache = {}
+local pngcachefiles = {}
+local function cachedpng(data)
+  local hash = md5.sumhexa(data)
+  local i = pngcache[hash]
+  if not i then
+    local path = ostmpname()
+    pngcachefiles[#pngcachefiles + 1] = path
+    open(path, "wb"):write(data):close()
+    -- local file = open(path, "wb"):write():close()
+    -- file:write(data)
+    -- file:close()
+  i = img.scan{filename = path}
+  pngcache[hash] = i
+end
+return i
+end
+
+local function get_png_glyph(gid, fontid, characters, haspng)
```

Comment:
Somehow the code from here on lost one level of indentation.

Reply:
Fixed